tests/functional/test_endpoint_rulesets.py

```diff
@@ -0,0 +1,312 @@
+                                ]
+                            },
+                            "headers": {},
+                        },
+                        "type": "endpoint",
+                    }
+                ],
+            }
+        )
+
+
+def iter_all_test_cases():
+    for service_name in ALL_SERVICES:
+        test_data = get_endpoint_tests_for_service(service_name)
+        for test_case in test_data['testCases']:
+            yield service_name, test_case
+
+
+def iter_provider_test_cases_that_produce(endpoints=False, errors=False):
+    for service_name, test in iter_all_test_cases():
+        input_params = test.get('params', {})
+        expected_object = test['expect']
+        if endpoints and 'endpoint' in expected_object:
+            yield service_name, input_params, expected_object['endpoint']
+        if errors and 'error' in expected_object:
+            yield service_name, input_params, expected_object['error']
+
+
+def iter_e2e_test_cases_that_produce(endpoints=False, errors=False):
+    for service_name, test in iter_all_test_cases():
+        # Not all test cases contain operation inputs for end-to-end tests.
+        if 'operationInputs' not in test:
+            continue
+        # Each test case can contain a list of input sets for the same
+        # expected result.
+        for op_inputs in test['operationInputs']:
+            op_params = op_inputs.get('operationParams', {})
+            # Test cases that use invalid bucket names as inputs fail in
+            # botocore because botocore validated bucket names before running
+            # endpoint resolution.
+            if op_params.get('Bucket') in ['bucket name', 'example.com#']:
+                continue
+            op_name = op_inputs['operationName']
+            builtins = op_inputs.get('builtInParams', {})
+
+            expected_object = test['expect']
+            if endpoints and 'endpoint' in expected_object:
+                expected_endpoint = expected_object['endpoint']
+                expected_props = expected_endpoint.get('properties', {})
+                expected_authschemes = [
+                    auth_scheme['name']
+                    for auth_scheme in expected_props.get('authSchemes', [])
+                ]
+                yield pytest.param(
+                    service_name,
+                    op_name,
+                    op_params,
+                    builtins,
+                    expected_endpoint,
+                    marks=pytest.mark.skipif(
+                        'sigv4a' in expected_authschemes,
+                        reason="Test case expects sigv4a which required CRT",
+                    ),
+                )
+            if errors and 'error' in expected_object:
+                yield pytest.param(
+                    service_name,
+                    op_name,
+                    op_params,
+                    builtins,
+                    expected_object['error'],
+                )
```

Comment:
iirc @nateprewitt suggested on my provider PR to have two variables, one for endpoint test cases and one for error test cases from the equivalent function there. That way, we only need to iterate over all of them once. Maybe we should do something similar here instead?